contracts/lottery/src/lib.rs

```diff
@@ -0,0 +1,609 @@
+/// 
+/// # Arguments
+/// 
+/// - `winners` - A map containing winning players with a number of correctly selected numbers
+/// - `thresholds` - Defined thresholds with prizes
+fn count_total_prizes_percentage(
+    winners: &Map<u32, Vec<Address>>,
+    thresholds: &Map<u32, u32>,
+) -> u32 {
+    winners
+        .iter()
+        .fold(0u32, |acc, (threshold_number, _)| {
+            let threshold_percentage = thresholds.get(threshold_number).unwrap();
+            let winners_count = winners.get(threshold_number).unwrap().len();
+            acc.add(threshold_percentage * winners_count)
+        })
+}
+
+/// Recalculates new thresholds in case total prizes percentage is above 100%
+///
+/// # Arguments
+/// 
+/// - `winners` - A map containing winning players with a number of correctly selected numbers
+/// - `thresholds` - Defined thresholds with prizes, will be updated
+/// in case total_prizes_percentage > 100
+/// - `total_prizes_percentage` - Total prizes percentage
+fn recalculate_new_thresholds(
+    winners: &Map<u32, Vec<Address>>,
+    thresholds: &mut Map<u32, u32>,
+    total_prizes_percentage: u32,
+) {
+    if total_prizes_percentage > 100 {
+        for threshold_number in thresholds.keys() {
+            if winners.contains_key(threshold_number) {
+                let winners_count = winners.get(threshold_number).unwrap().len();
+                let threshold_precentage = thresholds.get(threshold_number).unwrap();
+                let val =
+                    winners_count * threshold_precentage * 100 / total_prizes_percentage;
+                thresholds.set(threshold_number, val / winners_count);
+            } else {
+                thresholds.remove(threshold_number);
+            }
+        }
+    }
+}
+
+/// Counts properly selected numbers for a given ticket
+///
+/// # Returns
+/// 
+/// Counted matches
+/// 
+/// # Arguments
+/// 
+/// - `drawn_numbers` - An array containing numbers selected in this lottery,
+/// - `player_ticket` - Numbers selected by a player
+fn count_matches(drawn_numbers: &Vec<u32>, player_ticket: &LotteryTicket) -> u32 {
+    drawn_numbers
+        .iter()
+        .filter(|x| player_ticket.contains(x))
+        .count() as u32
+}
+
+/// Randomly draw numbers within a given range (1, max_range).
+/// Ensures that there are no duplicates
+///
+/// # Returns
+/// 
+/// A list of randomly selected number
+/// 
+/// # Arguments
+///
+/// - `env` - The environment for this contract.
+/// - `max_range` - Right boundary of the range players will select numbers from (1, max_range)
+/// - `number_of_numbers` - Number of numbers possible to select by players
+/// - `random_seed` - A seed provided by the admin, that will be combined with other environment elements
+fn draw_numbers(env: &Env, max_range: u32, number_of_numbers: u32, random_seed: u64) -> Vec<u32> {
+    let mut numbers = Vec::new(&env);
+    for n in 0..number_of_numbers {
+        let new_seed = random_seed + n as u64;
```

Comment:
We could make it non deterministic for production, but deterministic for tests. If we add the current timestamp as explained here ☝️ , then in tests, we can mock the time similar like we did [here](https://github.com/eigerco/nebula/blob/main/contracts/voting/src/test.rs#L147-L151) (maybe its time to move that test helpers functions to a common crate).

Well, we are working here with the assumption that ledger time would be difficult to predict (network calls, runtime, etc..). But maybe more investigation would be good, as i think is second precision.

Let me know your thoughts, maybe im wrong :)

Reply:
That would affect integration tests too. You can mock unit tests, but no the integration ones. I would rather not change that.

Reply:
Thats an option. But if we are already cheking behaviour in unit tests, i dont see the need of doing the same in integration tests. I would probably change the target of integration tests, like, checking theres a winner instead of an specific winner. I think the target with integration tests is to detect weird things like the one you detected in [#41](https://github.com/eigerco/nebula/issues/41) :)

Reply:
The idea of integration test is to check the full workflow on a specific network to be sure everything works as expected there. Let leave it as it is now and I'll think later how we could improve this.